R/driver-sql-server.R

```diff
@@ -73,11 +73,11 @@ setMethod("dbExistsTable", c("Microsoft SQL Server", "character"),
     stopifnot(length(name) == 1)
     if (isTempTable(conn, name, ...)) {
       query <- paste0("SELECT OBJECT_ID('tempdb..", name, "')")
```

Comment:
I wonder if it's worth checking to see if this works for synonyms too? (without the `tempdb..`). If so, that would substantially simplify this function, and maybe make it simpler?

Reply:
@simonpcouch Did you see this comment?

Reply:
I did! Just haven't made a moment to come back to this PR to address this comment and run some quick benchmarks to see how it affects performance. Will re-request review then, likely tomorrow!

Reply:
Oh sorry, I though you had re-requested a review, but I must've been misreading my notifications.

Reply:
Was almost able to make this work with:

``` r
    dots <- list(...)
    name_loc <- paste0(
      c(dots[["catalog_name"]], dots[["schema_name"]], name),
      collapse = "."
    )
    return(!is.na(dbGetQuery(conn, paste0("SELECT OBJECT_ID('", name_loc, "')"))[[1]]))
```

The only issue is that `OBJECT_ID()` seems to only be able to work with fully qualified names. i.e. 

``` r
library(DBI)
library(odbc)
con <- dbConnect(odbc(), dsn = "MicrosoftSQLServer", uid = "SA",
                 pwd = Sys.getenv("sqlServerPass"))

dbExecute(con, "create schema odbc")
#> [1] 0
dbExecute(con, "create table odbc.test (x int)")
#> [1] 0
dbExecute(con, "create synonym odbc.test2 for odbc.test")
#> [1] 0

# when fully qualified, works fine:
dbExistsTable(con, SQL("master.odbc.test2"))
#> [1] TRUE

# but not without schema/catalog specified
"test2" %in% dbListTables(con)
#> [1] TRUE
dbExistsTable(con, "test2")
#> [1] FALSE
```

<sup>Created on 2024-04-10 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Specifically, this triggered [this DBItest test](https://github.com/r-dbi/DBItest/blob/adc3f7f69de7addb96554f40ea581d0ac7c71279/R/spec-sql-exists-table.R#L92).

Reply:
Hmmmm, is `dbListTables()` correct here? i.e. `SELECT * FROM test2` won't work (IIUC) because it's not in the current schema. So should it actually be listed?

Reply:
:thinkies: The DBI docs for the generic read:

> dbListTables() returns a character vector that enumerates all tables and views in the database.

so... I think it's correct that `dbListTables()` would list `test2`. 

---

Some statements:

1) DBI docs: _dbListTables() returns a character vector that enumerates all tables and views in the database._
2) DBItest, linked above: _For all tables listed by `dbListTables()`, `dbExistsTable()` should return `TRUE`._
3) Us, on slack: _maybe the principle is that if `SELECT * FROM foo` works  then `dbExistsTable(con, "foo")` should return TRUE?_

So `dbListTables()` and `dbExistsTable()` should agree, and `dbListTables()` seems to be doing the right thing. The connection to `SELECT * FROM foo` doesn't seem to hold up, at least for SQL Server.

FWIW, with this PR:

``` r
library(DBI)
library(odbc)
con <- dbConnect(odbc(), dsn = "MicrosoftSQLServer", uid = "SA",
                 pwd = Sys.getenv("sqlServerPass"))


"test" %in% dbListTables(con)
#> [1] TRUE
dbExistsTable(con, "test")
#> [1] TRUE
dbGetQuery(con, "SELECT * FROM test")
#> Error in eval(expr, envir, enclos): nanodbc/nanodbc.cpp:1711: 00000
#> [Microsoft][ODBC Driver 18 for SQL Server][SQL Server]Invalid object name 'test'.

"test2" %in% dbListTables(con)
#> [1] TRUE
dbExistsTable(con, "test2")
#> [1] TRUE
dbGetQuery(con, "SELECT * FROM test2")
#> Error in eval(expr, envir, enclos): nanodbc/nanodbc.cpp:1711: 00000
#> [Microsoft][ODBC Driver 18 for SQL Server][SQL Server]Invalid object name 'test2'.
```

<sup>Created on 2024-04-10 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

With CRAN, both `"test2"` calls are `FALSE`.

Reply:
I'm pretty sure that the DBI docs are underspecified here — I think they really should read a "character vector that enumerates all tables and views in the database _for the active schema/catalog_.". 

So maybe we're blowing out the scope of this PR, but I think we should narrow down this behaviour because I think that we should be returning `FALSE` for the `test` calls too.